interfaces/cython/cantera/test/test_thermo.py

```diff
@@ -422,6 +422,125 @@ def test_equil_results(gas, fuel, ox, Y_Cf, Y_Of, Y_Co, Y_Oo, basis):
+            elif basis == "mole" and fraction_type == "oxidizer":
+                self.assertNear(gas["H2"].X[0], 0.1 * 4)
+                self.assertNear(gas["O2"].X[0], 0.1)
+                self.assertNear(gas["CO2"].X[0], 1 - 0.1 - 0.1 * 4)
+            elif basis == "mass" and fraction_type == "oxidizer":
+                self.assertNear(gas["O2"].Y[0], 0.1)
+                self.assertNear(gas["H2"].Y[0] / gas["O2"].Y[0], 4 * M_H2 / M_O2)
+
+            state = gas.state
+            self.assertNear(gas.equivalence_ratio("H2", "O2", include_species=["H2", "O2"], basis=basis), 2)
+            self.assertNear(gas.equivalence_ratio(include_species=["H2", "O2"], basis=basis), 2)
+            self.assertArrayNear(state, gas.state)
+
+        # brute force all possible input combinations
+        test_simple_dilution("diluent:0.3",   "mole")
+        test_simple_dilution({"diluent":0.3}, "mole")
+        test_simple_dilution("diluent:0.3",   "mass")
+        test_simple_dilution({"diluent":0.3}, "mass")
+
+        test_simple_dilution("fuel:0.1",   "mole")
+        test_simple_dilution({"fuel":0.1}, "mole")
+        test_simple_dilution("fuel:0.1",   "mass")
+        test_simple_dilution({"fuel":0.1}, "mass")
+
+        test_simple_dilution("oxidizer:0.1",    "mole")
+        test_simple_dilution({"oxidizer":0.1}, "mole")
+        test_simple_dilution("oxidizer:0.1",   "mass")
+        test_simple_dilution({"oxidizer":0.1}, "mass")
+
+    def test_equivalence_ratio_arbitrary_dilution(self):
+        fuel = "CH4:1,O2:0.01,N2:0.1,CO:0.05,CO2:0.02"
+        oxidizer = "O2:0.8,N2:0.2,CO:0.01,CH4:0.005,CO2:0.03"
+        diluent = "CO2:1,CO:0.025,N2:0.3,CH4:0.07,O2:0.09"
+        gas = ct.Solution("gri30.yaml")
+
+        gas.X = fuel
+        X_fuel = gas.X
+        gas.Y = fuel
+        Y_fuel = gas.Y
+        gas.X = oxidizer
+        X_oxidizer = gas.X
+        gas.Y = oxidizer
+        Y_oxidizer = gas.Y
+        gas.X = diluent
+        X_diluent = gas.X
+        gas.Y = diluent
+        Y_diluent = gas.Y
+
+        gas.set_equivalence_ratio(2, fuel, oxidizer)
+        X_Mix = gas.X
+        gas.set_equivalence_ratio(2, fuel,oxidizer, fraction="diluent:0.6", diluent=diluent)
+        X_expected = X_Mix * 0.4 + 0.6 * X_diluent
+        self.assertArrayNear(gas.X, X_expected)
+
+        gas.set_equivalence_ratio(2, fuel, oxidizer, basis="mass")
+        Y_Mix = gas.Y
+        gas.set_equivalence_ratio(2, fuel, oxidizer, fraction="diluent:0.6", diluent=diluent, basis="mass")
+        Y_expected = Y_Mix * 0.4 + 0.6 * Y_diluent
+        self.assertArrayNear(gas.Y, Y_expected)
+
+        gas.set_equivalence_ratio(0.8, fuel, oxidizer, basis="mass")
+        AFR = gas.stoich_air_fuel_ratio(fuel, oxidizer, basis="mass") / 0.8
+        gas.set_equivalence_ratio(0.8, fuel, oxidizer, fraction="fuel:0.05", diluent=diluent, basis="mass")
+        Y_expected = 0.05 * (Y_fuel + Y_oxidizer*AFR) + Y_diluent * (1 - (0.05 + 0.05 * AFR))
```

Comment:
```suggestion
        Y_expected = 0.05 * (Y_fuel + AFR * Y_oxidizer) + (1 - 0.05 * (1 + AFR)) * Y_diluent
```
is a minor simplification (same for `X` version)